File: matlab/EFA.m
classdef EFA
    %EFA - Evolving Factor Analysis
    
    properties
        I
        sigma = []
        svd_mode(1,:) char {mustBeMember(svd_mode,{'exact','fast'})} = 'exact'
    end
    properties(Dependent = true)
        Nr
        Nc
    end
    
    methods
        function obj = EFA(I,sigma,varargin)
            % EFA
            
            % assign input arguments
            obj.I = I;
            obj.sigma = sigma;
            
            % assign name, value pairs
            for j=1:2:(numel(varargin)-1)
                obj.(varargin{j}) = varargin{j+1};
            end
        end
        
        function val = get.Nr(obj)
            val = size(obj.I,1);
        end
        
        function val = get.Nc(obj)
            val = size(obj.I,2);
        end
        
        function sv = evolvingFactors(obj,k,direction,skip)
            %evolving_factors
            
            Ncols = obj.Nc;
            
            if nargin < 3 || isempty(direction)
                direction = 'forward';
            end
            
            if nargin < 4 || isempty(skip)
                skip = 1;
            end
            
            sv = NaN*ones(k,Ncols);
            
            for j=1:skip:Ncols
                switch lower(direction)
                    case 'forward'
                        cols = 1:j;
                    case {'reverse'}
                        cols = j:Ncols;
                end
                sv(:,j) = obj.svd(k,cols);
            end
            
        end
        
        function sv = svd(obj,k,cols,normalize,subtract)
            
            if nargin < 2 || isempty(k)
                k = obj.Nc;
            end
            if nargin < 3 || isempty(cols)
                cols = 1:obj.Nc;
            end
            if nargin < 4 || isempty(normalize)
                normalize = true;
            end
            if nargin < 5 || isempty(subtract)
                subtract = true;
            end
            
            sv = NaN*ones(k,1);
            
            % prepare matrix
            A = obj.I(:,cols);
            
            % apply sigma weight
            if ~isempty(obj.sigma) && size(obj.sigma,1) == obj.Nr
                if size(obj.sigma,2) == 1
                    A = A./obj.sigma;
                else
                    A = A./mean(obj.sigma(:,cols),2);
                end
            end
            
            if normalize
                A = A/sqrt(obj.Nr);
            end
            
            % calculate svd
            switch obj.svd_mode
                case 'exact'
                    [~,s,~] = svd(A,'econ');
                    s = diag(s);
                    s = s(1:min(k,numel(s)));
                case 'fast'
                    s = svds(A,k,'largest','Tolerance',1);
            end
            
            % subtract maximum s.v. of Gaussian random matrix
            if subtract && normalize
               s = s - sqrt(numel(cols)/obj.Nr);
            elseif subtract % && ~normalize
               s = s - sqrt(numel(cols));
            end
            
            sv(1:numel(s)) = s;
            
        end
        
        function [y,c,R] = quickRotate(obj,xstart,xend)
            ncomp = numel(xstart);
            
            % do SVD
            w = 1./mean(obj.sigma,2);
            [u,s,v] = svd(obj.I.*w,'econ');
            
            % get the important components
            u = u(:,1:ncomp);
            s = s(1:ncomp,1:ncomp);
            v = v(:,1:ncomp);
            
            % calculate optimal rotation
            R = zeros(ncomp,ncomp);
            
            for n=1:ncomp
                m = false(size(v,1),1);
                m(ceil(xstart(n)):floor(xend(n))) = true;

                vIn = v(m,:); % concentration basis inside peak
                vOut = v(not(m),:); % concentration basis outside peak
                
                % set up least squares problem
                A = vOut;
                B = mean(vIn,1);
                AA = A'*A;
                BB = B'*B;
                
                % large Lagrange multiplier enforces peak normalization
                lambda = 1E6*trace(AA)/trace(BB); 
                
                % solve the normal equations
                R(:,n) = (AA + lambda*(BB))\(lambda*B'*1);
            end

            % calculate I = y*c
            y = (u*s*pinv(R'))./w;
            c = R'*v';
        end
    end
    
    methods(Static = true)
        
        function [xinfl,xc,slope] = fitInflection(s,direction,threshold,window)
            N = size(s,1);
            xinfl = NaN*ones(N,1);
            xc = NaN*ones(N,1);
            slope = NaN*ones(N,1);
            x = 1:size(s,2);
            for j=1:N
                v = s(j,:);
                isIncl = ~isnan(v);
                xj = x(isIncl);
                yj = v(isIncl);
                switch lower(direction)
                    case 'forward'
                        x0 = xj(find(yj>threshold,1,'first'));
                    case 'reverse'
                        x0 = xj(find(yj>threshold,1,'last'));
                    otherwise
                        error('unexpected direction');
                end
                if isempty(x0)
                    continue;
                end
                xc(j) = x0;
                isWindow = (x0 - window/2) <= xj & (x0 + window/2) >= xj;
                npts = nnz(isWindow);
                if npts < 2
                    continue;
                end
                xw = xj(isWindow);
                yw = yj(isWindow);
                A = ones(npts,2);
                A(:,1) = xw(:) - x0;
                c = A\yw(:);
                xinfl(j) = x0 + (1-c(2))/c(1);
                slope(j) = c(1);
            end
            xinfl(xinfl<x(1)) = x(1);
            xinfl(xinfl>x(end)) = x(end);
            
        end
    end
end

File: python/efa.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import numpy as np
from scipy.sparse.linalg import svds
from scipy.sparse import csr_matrix

class efa:
    
    def __init__(self, I, sigma = np.array([])):
        self.I = I
        self.sigma = sigma
    
    @property
    def Nr(self):
        return self.I.shape[0]
    
    @property
    def Nc(self):
        return self.I.shape[1]
    
    def evolving_factors(self, k, direction = 'forward', skip = 1):
        
        Ncols = self.Nc
        
        sv = np.nan * np.ones((k, Ncols))
        
        for j in range(0,Ncols,skip):
            if direction.lower() == 'forward':
                cols = np.arange(j+1)
            elif direction.lower() == 'reverse':
                cols = np.arange(j,Ncols)
            sv[:,j] = self.svd(k,cols)
        
        return sv
        
    def svd(self, k = None, cols = np.array([]), normalize = True, subtract = True):
        
        if k == None:
            k = self.Nc
        
        if len(cols) == 0:
            cols = np.arange(self.Nc)
        
        sv = np.nan*np.ones(k);
        
        A = self.I[:,cols]
        
        if len(self.sigma) != 0 and self.sigma.shape[0] == self.Nr:
            if len(self.sigma.shape) == 1 or self.sigma.shape[1] == 1:
                A = np.diag(1/self.sigma) @ A
            else:
                A = np.diag(1/np.mean(self.sigma[:,cols],1)) @ A
        
        if normalize:
            A = A / np.sqrt(self.Nr)
        
        s = np.linalg.svd(A, full_matrices=False, compute_uv=False)
        s = s[:min(k,len(s))]
        
        if subtract and normalize:
            s = s - np.sqrt(cols.shape[0]/self.Nr)
        elif subtract:
            s = s - np.sqrt(cols.shape[0])
        
        sv[:len(s)] = s
        return sv
    
    def quick_rotate(self, xstart, xend):
        ncomp = len(xstart)
        
        w = 1 / np.mean(self.sigma,1)
        [u, s, v] = np.linalg.svd(np.diag(w,0) @ self.I, full_matrices=False)
        
        u = u[:,:ncomp]
        s = s[:ncomp]
        v = v.T[:,:ncomp]
        
        R = np.zeros((ncomp,ncomp))
        
        for n in range(ncomp):
            m = np.full(v.shape[0],False)
            m[int(np.ceil(xstart[n])):int(np.floor(xend[n]))] = True
            
            v_in = v[m,:]
            v_out = v[~m,:]
            
            A = v_out
            B = np.mean(v_in, 0)
            AA = A.T @ A
            BB = np.outer(B,B)
            
            lmbd = 1E6 * np.trace(AA) / np.trace(BB)
            
            R[:,n] = np.linalg.solve(AA + lmbd * BB, lmbd * B * 1)
        
        y = (u @ np.diag(s) @ np.linalg.pinv(R.T)) / w[:,np.newaxis]
        c = R.T @ v.T
        
        return [y, c, R]
    
    @staticmethod
    def fit_inflection(s, direction, threshold, window):
        N = s.shape[0]
        xinfl = np.nan * np.ones(N)
        xc = np.nan *np.ones(N)
        slope = np.nan *np.ones(N)
        x = np.arange(s.shape[1])
        
        for j in range(N):
            v = s[j, :]
            is_incl = ~np.isnan(v)
            xj = x[is_incl]
            yj = v[is_incl]
            pos = np.argwhere(yj>threshold)
            if len(pos) != 0:
                if direction == 'forward':
                    x0 = xj[pos[0]]
                elif direction == 'reverse':
                    x0 = xj[pos[-1]]
                else:
                    raise ValueError('unexpected direction')
            else:
                continue
            xc[j] = x0
            is_window = np.logical_and((x0 - window / 2) <= xj, (x0 + window / 2) >= xj)
            npts = np.count_nonzero(is_window)
            if npts < 2:
                continue
            xw = xj[is_window]
            yw = yj[is_window]
            A = np.ones((npts,2))
            A[:, 0] = xw - x0
            c = np.linalg.lstsq(A, yw, rcond=None)[0]
            xinfl[j] = x0 + (1 - c[1]) / c[0]
            slope[j] = c[0]
        
        xinfl[~np.isnan(xinfl)][xinfl[~np.isnan(xinfl)]<x[0]] = x[0]
        xinfl[~np.isnan(xinfl)][xinfl[~np.isnan(xinfl)]>x[-1]] = x[-1]
        
        return [xinfl, xc, slope]
        